contrib/http_base/src/main/java/io/opentelemetry/contrib/http/core/HttpClientHandler.java

```diff
@@ -0,0 +1,183 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.contrib.http.core;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static io.opentelemetry.contrib.http.core.HttpTraceConstants.INSTRUMENTATION_LIB_ID;
+
+import io.opentelemetry.OpenTelemetry;
+import io.opentelemetry.context.propagation.HttpTextFormat;
+import io.opentelemetry.distributedcontext.DistributedContext;
+import io.opentelemetry.distributedcontext.DistributedContextManager;
+import io.opentelemetry.distributedcontext.Entry;
+import io.opentelemetry.distributedcontext.EntryMetadata.EntryTtl;
+import io.opentelemetry.metrics.Meter;
+import io.opentelemetry.trace.Span;
+import io.opentelemetry.trace.Span.Kind;
+import io.opentelemetry.trace.SpanContext;
+import io.opentelemetry.trace.Tracer;
+import java.util.logging.Logger;
+import javax.annotation.Nullable;
+
+/**
+ * This helper class provides routine methods to instrument HTTP servers.
+ *
+ * @param <Q> the HTTP request entity.
+ * @param <P> the HTTP response entity.
+ * @param <C> the type of the carrier.
+ */
+public class HttpClientHandler<Q, P, C> extends AbstractHttpHandler<Q, P> {
+
+  private static final Logger LOGGER = Logger.getLogger(HttpClientHandler.class.getName());
+
+  private final Tracer tracer;
+  private final DistributedContextManager contextManager;
+  private final HttpTextFormat.Setter<C> setter;
+  private final HttpTextFormat<SpanContext> spanContextHttpTextFormat;
+  private final HttpTextFormat<DistributedContext> distributedContextHttpTextFormat;
+
+  /**
+   * Constructs a handler object.
+   *
+   * @param extractor used to extract information from request/response
+   * @param setter the setter used to extract information from the carrier
+   */
+  public HttpClientHandler(HttpExtractor<Q, P> extractor, HttpTextFormat.Setter<C> setter) {
+    this(extractor, setter, null, null, null, null);
+  }
+
+  /**
+   * Constructs a handler object.
+   *
+   * @param extractor used to extract information from request/response
+   * @param setter the setter used to extract information from the carrier
+   * @param statusConverter the HTTP status to OT status translator
+   * @param tracer the OT tracer
+   * @param contextManager the OT distributed context manager
+   * @param meter the OT meter
+   */
+  public HttpClientHandler(
+      HttpExtractor<Q, P> extractor,
+      HttpTextFormat.Setter<C> setter,
+      StatusCodeConverter statusConverter,
+      Tracer tracer,
+      DistributedContextManager contextManager,
+      Meter meter) {
+    super(extractor, statusConverter, meter);
```

Comment:
I'll leave my general comment here. I don't think there should be a class hierarchy here at all. If we need some common functions, put them in a class that both client and server code can access statically, rather than introducing a class hierarchy.

Reply:
Or, maybe not statically, come to think of it. We should favor composition. Give this class access to the functions via an instance field, rather than via a class hierarchy.

Reply:
Static will not work. OpenCensus uses this approach. Based on how it is used there, these classes are never subclassed again

Reply:
I feel pretty strongly that introducing a class hierarchy here is the wrong design, even if it was used in open census. I'd very much prefer that this be implemented via composition, rather than inheritance.

Reply:
I looked at doing this. Some of the code ends up complicated and contrived. It would make more sense to copy and paste rather than use composition. Let me know if you want me to do that.

Reply:
Let's keep it as-is for now, since you've got a bunch of instrumentation queued up using this setup already. I'll look into seeing what it would take to use composition after we've got it all working. :)